kafka-producer/src/main/java/com/expedia/www/haystack/pipes/kafka/producer/KafkaToKafkaPipeline.java

```diff
@@ -0,0 +1,118 @@
+    @VisibleForTesting
+    static Counter kafkaProducerCounter;
+    static Counter requestCounter;
+    static Timer kafkaProducerTimer;
+    private final TagFlattener tagFlattener = new TagFlattener();
+    private final Map<SpanKeyExtractor, List<KafkaProducer<String, String>>> kafkaProducerMap;
+
+    public KafkaToKafkaPipeline(MetricRegistry metricRegistry,
+                                Map<SpanKeyExtractor, List<KafkaProducer<String, String>>> kafkaProducerMap) {
+        this.kafkaProducerMap = kafkaProducerMap;
+        requestCounter = metricRegistry.counter("REQUEST");
+        kafkaProducerTimer = metricRegistry.timer("KAFKA_PRODUCER_POST_TIMER");
+        kafkaProducerCounter = metricRegistry.counter("KAFKA_PRODUCER_POST_COUNTER");
+    }
+
+    @Override
+    public void apply(String key, Span value) {
+        requestCounter.inc();
+        Timer.Context time = kafkaProducerTimer.time();
+        kafkaProducerMap.forEach((spanKeyExtractor, kafkaProducers) -> {
+            List<String> kafkaTopics = new ArrayList<>();
+            final String kafkaKey = spanKeyExtractor.getKey();
+            Optional<String> optionalMsg = spanKeyExtractor.extract(value);
+            if (!optionalMsg.isPresent()) {
+                logger.debug("Extractor skipped the span: {}", value);
+                return;
+            }
+            String kafkaMsg = optionalMsg.get();
+            kafkaTopics.addAll(spanKeyExtractor.getTopics());
+            String msgWithFlattenedTags = tagFlattener.flattenTags(kafkaMsg);
+            logger.info("KafkaProducer sending message: {},with key: {}  ", msgWithFlattenedTags, kafkaKey);
+            kafkaProducers.forEach(kafkaProducer -> produceToKafkaTopics(kafkaProducer, kafkaTopics, kafkaKey, msgWithFlattenedTags));
+
+        });
+        time.stop();
+    }
+
+    public void produceToKafkaTopics(KafkaProducer<String, String> kafkaProducer, List<String> kafkaTopics, String kafkaKey, String jsonWithFlattenedTags) {
+        kafkaTopics.forEach(topic -> {
+            final ProducerRecord<String, String> producerRecord =
+                    factory.createProducerRecord(topic, kafkaKey, jsonWithFlattenedTags);
+
+            final KafkaCallback callback; // callback must returnObject()
+            try {
+                callback = OBJECT_POOL.borrowObject();
+                kafkaProducerCounter.inc();
+                // TODO Put the Span value into the callback so that it can write it to Kafka for retry
+                kafkaProducer.send(producerRecord, callback);
+                logger.info("Kafka message sent on topic: {}", topic);
+            } catch (Exception exception) {
+                // Must format below because log4j2 underneath slf4j doesn't handle .error(varargs) properly
+                final String message = String.format(ERROR_MSG, jsonWithFlattenedTags, exception.getMessage());
+                logger.error(message, exception);
```

Comment:
We would need a metric here to help us know if there is any issue. And preferably, separate metric for each producer

Reply:
👍